genie-common-internal/src/main/java/com/netflix/genie/common/internal/services/impl/JobDirectoryManifestServiceImpl.java

```diff
@@ -0,0 +1,95 @@
+/*
+ *
+ *  Copyright 2019 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.common.internal.services.impl;
+
+import com.netflix.genie.common.internal.dto.DirectoryManifest;
+import com.netflix.genie.common.internal.services.JobDirectoryManifestService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.cache.annotation.Cacheable;
+
+import java.io.IOException;
+import java.nio.file.Path;
+
+/**
+ * Implementation of {@link JobDirectoryManifestService} that caches generated manifests to avoid re-calculating one
+ * for each request.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class JobDirectoryManifestServiceImpl implements JobDirectoryManifestService {
+
+    /**
+     * Name of the cache bean that this implementation relies on.
+     */
+    public static final String CACHE_NAME = "JobDirectoryManifestServiceImplCache";
+    private final DirectoryManifest.Factory factory;
+    private final boolean checksum;
+
+    /**
+     * Constructor.
+     *
+     * @param factory  the directory manifest factory
+     * @param checksum whether to calculate and include checksums entries
+     */
+    public JobDirectoryManifestServiceImpl(
+        final DirectoryManifest.Factory factory,
+        final boolean checksum
+    ) {
+        this.factory = factory;
+        this.checksum = checksum;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public DirectoryManifest getDirectoryManifest(final Path jobDirectoryPath) throws IOException {
+        // TODO: Introduce a timer
+        final long start = System.nanoTime();
+        try {
+            return getCachedDirectoryManifest(jobDirectoryPath);
+        } finally {
+            final long end = System.nanoTime();
+            log.debug(
+                "Producing manifest for {} took {}ns",
+                jobDirectoryPath.normalize().toAbsolutePath().toString(),
+                end - start
+            );
+        }
+    }
+
+    /**
+     * Internal cached method (cannot be private to work).
+     *
+     * @param jobDirectoryPath the job directory
+     * @return a cached manifest for the input path
+     * @throws IOException in case of error
+     */
+    @Cacheable(
```

Comment:
Might be better to put this on an interface?

Reply:
But then the service would need to be something like `CachingJobDirectoryManfestService`.
I intentionally left the caching behavior as implementation specific.

Reply:
I see